# AI Resume Analyzer & ATS Score Predictor

### End-to-End NLP + LLM Project

Author: Gopal Yadav

Technologies:
- Python
- NLP
- Sentence Transformers
- Gemini LLM
- Streamlit

# Phase 1: Importing Libraries

In [1]:
import pdfplumber
import pandas as pd
import numpy as np
import spacy

## Phase 2 : Resume Parsing

In [2]:
def extract_text_from_pdf(pdf_path):

    try:

        full_text = ""

        with pdfplumber.open(pdf_path) as pdf:

            for page in pdf.pages:

                text = page.extract_text()

                if text:

                    full_text += text + "\n"

        return full_text

    except Exception as e:

        print(f"Error: {e}")

        return None

In [3]:
# Testing the function :
resume_text = extract_text_from_pdf("sample_resume.pdf")
print(resume_text)

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI),MachineLearning(ML),DeepLearning,NaturalLanguageProcessing(NLP)

In [4]:
# Test 1 (Correct File)
resume_text = extract_text_from_pdf("sample_resume.pdf")

print(resume_text)

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI),MachineLearning(ML),DeepLearning,NaturalLanguageProcessing(NLP)

In [5]:
# Test 1 (Wrong File)
wrong_resume = extract_text_from_pdf("abc.pdf")

Error: [Errno 2] No such file or directory: 'abc.pdf'


In [6]:
clean_text = "I am studying machinelearning and deeplearning."
# Normalize common merged technical terms

clean_text = clean_text.replace("machinelearning", "machine learning")
clean_text = clean_text.replace("deeplearning", "deep learning")
clean_text = clean_text.replace("naturallanguageprocessing", "natural language processing")
clean_text = clean_text.replace("scikitlearn", "scikit-learn")
clean_text = clean_text.replace("computervision", "computer vision")
clean_text = clean_text.replace("promptengineering", "prompt engineering")
clean_text = clean_text.replace("powerbi", "power bi")
clean_text = clean_text.replace("githubcom", "github")

print(clean_text)

I am studying machine learning and deep learning.


# Phase 3: Text Preprocessing

Convert Text to Lowercase

In [7]:
print("Python" == "python")

False


In [8]:
print(resume_text)

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI),MachineLearning(ML),DeepLearning,NaturalLanguageProcessing(NLP)

In [9]:
resume_text = extract_text_from_pdf("sample_resume.pdf")

print(resume_text)

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI),MachineLearning(ML),DeepLearning,NaturalLanguageProcessing(NLP)

In [10]:
resume_text = extract_text_from_pdf("sample_resume.pdf")

In [11]:
clean_text = resume_text.lower()

print(clean_text)

gopal yadav
mumbai,maharashtra,india | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
aboutme
final-yearb.sc. computersciencestudentspecializinginmachinelearning,deeplearning,nlp,andgenerativeai,withthree
end-to-endaiprojectsindependentlybuilt,evaluated,anddeployed. deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. experiencedacrossthefullmllifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingpython,scikit-learn,tensorflow/keras,langchain,andibmwatsonx. completed8industrycertificationsfromibm,aws,and
microsoft/linkedincoveringmachinelearning,dataanalysis,sql,andgenerativeai,reflectingconsistentself-directedupskilling.
technicalskills
programming&querylanguages: python,sql
coredomains: artificialintelligence(ai),machinelearning(ml),deeplearning,naturallanguageprocessing(nlp)

Removing Extra Spaces

In [12]:
# See the Hidden Characters
print(repr(clean_text[:500]))

'gopal yadav\nmumbai,maharashtra,india | +91-7400124757 | gopalyadav18832@gmail.com\nlinkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai\naboutme\nfinal-yearb.sc. computersciencestudentspecializinginmachinelearning,deeplearning,nlp,andgenerativeai,withthree\nend-to-endaiprojectsindependentlybuilt,evaluated,anddeployed. deliveredmeasurablemodelperformance,includinga98.08%\naccuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. experiencedacrossthefullmllifecycle–data\npreprocessi'


In [13]:
# Clean Extra Spaces
clean_text = " ".join(clean_text.split())
print(clean_text[:1000])

gopal yadav mumbai,maharashtra,india | +91-7400124757 | gopalyadav18832@gmail.com linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai aboutme final-yearb.sc. computersciencestudentspecializinginmachinelearning,deeplearning,nlp,andgenerativeai,withthree end-to-endaiprojectsindependentlybuilt,evaluated,anddeployed. deliveredmeasurablemodelperformance,includinga98.08% accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. experiencedacrossthefullmllifecycle–data preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment– usingpython,scikit-learn,tensorflow/keras,langchain,andibmwatsonx. completed8industrycertificationsfromibm,aws,and microsoft/linkedincoveringmachinelearning,dataanalysis,sql,andgenerativeai,reflectingconsistentself-directedupskilling. technicalskills programming&querylanguages: python,sql coredomains: artificialintelligence(ai),machinelearning(ml),deeplearning,naturallanguageprocessing(nlp)

Remove Special Characters

In [14]:
import re

In [15]:
sample = "Python, SQL!!! TensorFlow (Keras)"
clean_sample = re.sub(r"[^\w\s]", "", sample)
print(clean_sample)

Python SQL TensorFlow Keras


In [16]:
# No special characters :
clean_text = re.sub(r"[^\w\s]", "", clean_text)
print(clean_text[:1000])

gopal yadav mumbaimaharashtraindia  917400124757  gopalyadav18832gmailcom linkedincomingopalyadavb637b73a0  githubcomgopalyadavai aboutme finalyearbsc computersciencestudentspecializinginmachinelearningdeeplearningnlpandgenerativeaiwiththree endtoendaiprojectsindependentlybuiltevaluatedanddeployed deliveredmeasurablemodelperformanceincludinga9808 accuracyhealthcareriskmodelandan80accuracysentimentanalysismodel experiencedacrossthefullmllifecycledata preprocessingexploratorydataanalysisfeatureengineeringmodelbuildinghyperparametertuningevaluationanddeployment usingpythonscikitlearntensorflowkeraslangchainandibmwatsonx completed8industrycertificationsfromibmawsand microsoftlinkedincoveringmachinelearningdataanalysissqlandgenerativeaireflectingconsistentselfdirectedupskilling technicalskills programmingquerylanguages pythonsql coredomains artificialintelligenceaimachinelearningmldeeplearningnaturallanguageprocessingnlpcomputer visionlargelanguagemodelsllmsgenerativeaipromptengineeringretr

Tokenization

In [17]:
tokens = clean_text.split()
print(tokens[:100])

['gopal', 'yadav', 'mumbaimaharashtraindia', '917400124757', 'gopalyadav18832gmailcom', 'linkedincomingopalyadavb637b73a0', 'githubcomgopalyadavai', 'aboutme', 'finalyearbsc', 'computersciencestudentspecializinginmachinelearningdeeplearningnlpandgenerativeaiwiththree', 'endtoendaiprojectsindependentlybuiltevaluatedanddeployed', 'deliveredmeasurablemodelperformanceincludinga9808', 'accuracyhealthcareriskmodelandan80accuracysentimentanalysismodel', 'experiencedacrossthefullmllifecycledata', 'preprocessingexploratorydataanalysisfeatureengineeringmodelbuildinghyperparametertuningevaluationanddeployment', 'usingpythonscikitlearntensorflowkeraslangchainandibmwatsonx', 'completed8industrycertificationsfromibmawsand', 'microsoftlinkedincoveringmachinelearningdataanalysissqlandgenerativeaireflectingconsistentselfdirectedupskilling', 'technicalskills', 'programmingquerylanguages', 'pythonsql', 'coredomains', 'artificialintelligenceaimachinelearningmldeeplearningnaturallanguageprocessingnlpcomput

In [18]:
print("Total Tokens:", len(tokens))

Total Tokens: 81


# Phase 4 : Skill Extraction

In [19]:
skills = [

    # Programming
    "python",
    "sql",
    "java",
    "c++",

    # Machine Learning
    "machine learning",
    "machinelearning",
    "deep learning",
    "deeplearning",

    # NLP
    "nlp",
    "natural language processing",
    "naturallanguageprocessing",
    "transformers",
    "bert",
    "llm",
    "rag",
    "langchain",

    # Computer Vision
    "computer vision",
    "computervision",
    "opencv",
    "cnn",

    # Libraries
    "tensorflow",
    "keras",
    "scikit-learn",
    "scikitlearn",
    "pandas",
    "numpy",
    "matplotlib",
    "seaborn",
    "xgboost",
    "adaboost",

    # BI
    "power bi",
    "powerbi",
    "tableau",

    # Deployment
    "streamlit",
    "fastapi",

    # Cloud
    "aws",
    "azure",
    "ibm watsonx",
    "watsonx",

    # Databases
    "mysql",
    "sqlite",

    # Version Control
    "git",
    "github"
]

In [20]:
# Extract Skills
found_skills = []

for skill in skills:
    if skill in clean_text:
        found_skills.append(skill)

found_skills = sorted(set(found_skills))

print(found_skills)

['adaboost', 'aws', 'computer vision', 'deeplearning', 'git', 'github', 'keras', 'langchain', 'llm', 'machinelearning', 'matplotlib', 'naturallanguageprocessing', 'nlp', 'numpy', 'pandas', 'power bi', 'python', 'rag', 'scikitlearn', 'seaborn', 'sql', 'sqlite', 'streamlit', 'tableau', 'tensorflow', 'watsonx', 'xgboost']


Skill Extraction using spaCy PhraseMatcher

In [21]:
# Import spaCy
import spacy
from spacy.matcher import PhraseMatcher

In [22]:
# Load the NLP Model
nlp = spacy.load("en_core_web_sm")
print("spaCy Loaded Successfully!")

spaCy Loaded Successfully!


In [23]:
# Creating PhraseMatcher
matcher = PhraseMatcher(nlp.vocab, attr="LOWER")

In [24]:
# Convert Skills into spaCy Patterns
patterns = [nlp.make_doc(skill) for skill in skills]

In [25]:
print(type(patterns[0]))

<class 'spacy.tokens.doc.Doc'>


In [26]:
# Add Skill Patterns to PhraseMatcher
matcher.add("SKILLS", patterns)
print("Skill patterns added successfully!")

Skill patterns added successfully!


In [27]:
# Finding Skill Matches in Resume

doc = nlp(clean_text)
matches = matcher(doc)
matched_skills = []
for match_id, start, end in matches:
    matched_skills.append(doc[start:end].text)
matched_skills = sorted(set(matched_skills))
print("Matched Skills:")
print(matched_skills)

Matched Skills:
[]


## Observation

The PhraseMatcher implementation was successfully verified using sample text.

However, the extracted resume PDF merges multiple words together (e.g., "usingpythonscikitlearntensorflow..."), causing spaCy to treat them as a single token.

As a result, token-based phrase matching cannot detect individual skills from this PDF. This is a limitation of the PDF text extraction process rather than the PhraseMatcher itself.

To overcome this limitation, the next implementation uses Sentence Transformers for semantic skill extraction.

In [28]:
# Semantic Skill Extraction using Sentence Transformers

# Instead of checking whether a skill exactly matches the text (like "python"), we want the model to understand meaning.
# What are Sentence Transformers?
# A Sentence Transformer is a pre-trained Deep Learning model that converts text into embeddings (vectors of numbers).

In [29]:
# Semantic Skill Extraction using Sentence Transformers

from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

In [30]:
# Load the model 

model = SentenceTransformer("all-MiniLM-L6-v2")                  
print("Sentence Transformer Loaded Successfully!")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

Sentence Transformer Loaded Successfully!


It is a pre-trained Sentence Transformer model from Hugging Face.

It is:
Small
Fast
Very accurate
Commonly used in semantic search applications

The first time you run it, it may take a minute or two because it downloads the model.

In [31]:
# Create Skill Embeddings

skill_embeddings = model.encode(skills)

print("Total Skill Embeddings:", len(skill_embeddings))

Total Skill Embeddings: 43


In [32]:
# Convert Resume into an Embedding

resume_embedding = model.encode([resume_text])
print(resume_embedding.shape)

(1, 384)


In [33]:
# Compare Resume with Each Skill

semantic_skills = []

for skill, embedding in zip(skills, skill_embeddings):

    similarity = cosine_similarity(
        [embedding],
        resume_embedding
    )[0][0]

    if similarity > 0.30:
        semantic_skills.append((skill, similarity))

In [34]:
# Sort the Results

semantic_skills = sorted(
    semantic_skills,
    key=lambda x: x[1],
    reverse=True
)

for skill, score in semantic_skills:
    print(f"{skill:30} {score:.3f}")

tensorflow                     0.380
machinelearning                0.371
deeplearning                   0.370
scikit-learn                   0.365
deep learning                  0.340
scikitlearn                    0.330
keras                          0.305
machine learning               0.300


# Phase 5 : Intelligent Semantic Skill Extraction

Split Resume into Sentences

In [35]:
sentences = []

doc = nlp(resume_text)

for sent in doc.sents:
    sentence = sent.text.strip()

    if len(sentence) > 5:
        sentences.append(sentence)

print("Total Sentences:", len(sentences))

Total Sentences: 19


In [36]:
for i, sentence in enumerate(sentences):
    print(f"{i+1}. {sentence}")

1. GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc.
2. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed.
3. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx.
4. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
5. TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI),MachineLearning(ML),DeepLearning,NaturalLanguage

Creating Embeddings for Every Sentence

In [37]:
sentence_embeddings = model.encode(sentences)
print("Total Embeddings:", len(sentence_embeddings))

Total Embeddings: 19


## Phase 5 Observation

Sentence Transformers were successfully integrated into the project.

An experiment was performed to semantically extract skills directly from the resume. However, due to the PDF extraction process merging multiple words (e.g., "usingpythonscikitlearntensorflow..."), semantic skill extraction produced incomplete results.

Instead of relying on semantic skill extraction, the project uses Sentence Transformers for Resume vs Job Description semantic similarity, which is more robust and closely matches how modern ATS systems evaluate resumes.

# Phase 6 : Resume vs Job Description Matching

In [38]:
# Resume vs Job Description Matching

# Read Job Description

with open("sample_jd.txt", "r", encoding="utf-8") as file:
    job_description = file.read()

print(job_description)

AI Engineer Intern

We are looking for an AI Engineer Intern with strong knowledge of Python, SQL, Machine Learning, Deep Learning, NLP, Computer Vision, LLMs, Generative AI, and RAG.

Required Skills:
- Python
- SQL
- Machine Learning
- Deep Learning
- TensorFlow
- Keras
- Scikit-learn
- Pandas
- NumPy
- NLP
- Computer Vision
- LLMs
- Prompt Engineering
- LangChain
- RAG
- Streamlit
- Git
- GitHub
- AWS

Responsibilities:
- Build end-to-end AI applications.
- Train and evaluate Machine Learning models.
- Develop NLP and Generative AI solutions.
- Deploy AI applications using Streamlit.
- Work with LLMs and Retrieval-Augmented Generation (RAG).
- Collaborate with AI Engineers.


Creating Job Description Embedding

In [39]:
jd_embedding = model.encode([job_description])

print(jd_embedding.shape)

(1, 384)


Creating Resume Embedding

In [40]:
resume_embedding = model.encode([resume_text])

print(resume_embedding.shape)

(1, 384)


Calculating Cosine Similarity

In [41]:
similarity = cosine_similarity(
    resume_embedding,
    jd_embedding
)[0][0]

print("Similarity Score:", similarity)

Similarity Score: 0.3753534


Convert to Resume Match Score

In [42]:
match_score = similarity * 100

print("=" * 50)
print(f"Resume Match Score : {match_score:.2f}%")
print("=" * 50)

Resume Match Score : 37.54%


# Phase 7 : ATS Score Calculation

Our ATS Formula

We'll use this weighted formula:

Component	Weight

Resume-JD Semantic Match	30

Skill Match	                40

Education	                10

Experience	                10

Resume Completeness	        10

Total	                    100


This is much closer to a real ATS than using semantic similarity alone.

In [43]:
# Extract Skills from Job Description

jd_doc = nlp(job_description)

In [44]:
matches = matcher(jd_doc)

jd_skills = []

for match_id, start, end in matches:
    skill = jd_doc[start:end].text.lower()

    if skill not in jd_skills:
        jd_skills.append(skill)

jd_skills = sorted(jd_skills)

print("Skills Found in Job Description:\n")

for skill in jd_skills:
    print(skill)

Skills Found in Job Description:

aws
computer vision
deep learning
git
github
keras
langchain
machine learning
nlp
numpy
pandas
python
rag
scikit-learn
sql
streamlit
tensorflow


In [45]:
# Compare Resume Skills with JD Skills

matched_skills = list(
    set(found_skills).intersection(set(jd_skills))
)

matched_skills = sorted(matched_skills)

print("Matched Skills:\n")

for skill in matched_skills:
    print(skill)

Matched Skills:

aws
computer vision
git
github
keras
langchain
nlp
numpy
pandas
python
rag
sql
streamlit
tensorflow


In [46]:
# Finding Missing Skills
missing_skills = list(
    set(jd_skills) - set(found_skills)
)

missing_skills = sorted(missing_skills)

print("Missing Skills:\n")

for skill in missing_skills:
    print(skill)

Missing Skills:

deep learning
machine learning
scikit-learn


In [47]:
# Calculate Skill Match Score
skill_match = (len(matched_skills) / len(jd_skills)) * 100

print("=" * 50)
print(f"Skill Match Score : {skill_match:.2f}%")
print("=" * 50)

Skill Match Score : 82.35%


# Phase 8 : ATS Score Calculation

In [48]:
# Semantic Score (30 Marks)

semantic_score = (match_score / 100) * 30
print(f"Semantic Score : {semantic_score:.2f}/30")

Semantic Score : 11.26/30


In [49]:
# Calculating Skill Score (40 Marks)

skills_score = (skill_match / 100) * 40

print(f"Skill Score : {skills_score:.2f}/40")

Skill Score : 32.94/40


In [50]:
# Education Score 

education_score = 10

print(f"Education Score : {education_score}/10")

Education Score : 10/10


In [51]:
# Experience Score
experience_score = 10

print(f"Experience Score : {experience_score}/10")

Experience Score : 10/10


In [52]:
# Resume Completeness 

completeness_score = 10

print(f"Resume Completeness : {completeness_score}/10")

Resume Completeness : 10/10


In [53]:
# Final ATS Score

ats_score = (
    semantic_score +
    skills_score +
    education_score +
    experience_score +
    completeness_score
)

print("=" * 60)
print(f"Final ATS Score : {ats_score:.2f}/100")
print("=" * 60)

Final ATS Score : 74.20/100


# Phase 9 : Google Gemini AI Resume Suggestionsv

Instead of showing only:

ATS Score : 74.20%

our application will also generate an intelligent report such as:

Resume Analysis

ATS Score : 74.20%

Strengths:
 Strong Python skills
 Good NLP knowledge
 TensorFlow experience
 Git & GitHub

Missing Skills:
• Machine Learning
• Deep Learning
• Scikit-learn

Suggestions:
1. Add Machine Learning explicitly in the Skills section.
2. Mention Deep Learning projects.
3. Include Scikit-learn in Technical Skills.
4. Quantify achievements using numbers.
5. Add more AI Engineer keywords.

In [54]:
# Install Google Gemini SDK

import sys
!{sys.executable} -m pip install -q google-generativeai

In [55]:
# Importing the Library

# import google.generativeai as genai

print("Google Gemini Imported Successfully!")

Google Gemini Imported Successfully!


In [96]:
# Configure Gemini

from google import genai

client = genai.Client(api_key="YOUR_API_KEY")

print("Gemini Configured Successfully!")

Gemini Configured Successfully!


In [57]:
# Creating the Gemini Model 

MODEL_NAME = "gemini-2.5-flash"
print("Gemini Model Ready!")

Gemini Model Ready!


In [58]:
# Create the AI Prompt

prompt = f"""
You are an expert ATS Resume Reviewer.

Analyze the following resume against the job description.

Resume Match Score:
{match_score:.2f}%

Skill Match Score:
{skill_match:.2f}%

ATS Score:
{ats_score:.2f}/100

Matched Skills:
{matched_skills}

Missing Skills:
{missing_skills}

Resume:
{resume_text}

Job Description:
{job_description}

Please provide:

1. Overall Resume Review
2. Resume Strengths
3. Missing Skills
4. Weaknesses
5. Suggestions to Improve ATS Score
6. Final Recommendation
"""

In [59]:
!pip install -U google-genai

In [61]:
# Generating the AI Suggestions

response = client.models.generate_content(
    model=MODEL_NAME,
    contents=prompt
)
print(response.text)

Direct use of automatic function calling (AFC) in Models.generate_content is not recommended. Instead, we recommend to use AFC in Chat.send_message. Similarly, direct use of AFC in Models.generate_content_stream is not recommended. Instead, we recommend to use AFC in Chat.send_message_stream.


This is a comprehensive ATS review of Gopal Yadav's resume against the AI Engineer Intern job description.

---

### 1. Overall Resume Review

Gopal Yadav's resume presents a strong profile for an AI Engineer Intern position, demonstrating solid foundational knowledge and hands-on experience in key AI/ML domains. The "About Me" section is particularly effective, summarizing his specialization, project experience, quantifiable achievements, and technical breadth. His project section is well-detailed, showcasing end-to-end development, specific technologies, and impressive metrics.

While the provided ATS Resume Match Score of 37.54% seems exceptionally low given the content, and the "Missing Skills" report is inaccurate based on a manual review, the resume *conceptually* aligns very well with the job description. The candidate explicitly mentions and demonstrates almost all the required skills. The discrepancies likely stem from how the ATS is configured (e.g., strict keyword matching, 

# Phase 10 : Dynamic ATS Scoring

In [62]:
education_keywords = [
    "bachelor",
    "b.sc",
    "btech",
    "b.tech",
    "computer science",
    "engineering",
    "master",
    "m.sc",
    "mtech"
]

education_score = 0

resume_lower = resume_text.lower()

for keyword in education_keywords:
    if keyword in resume_lower:
        education_score = 10
        break

print("=" * 50)
print(f"Education Score : {education_score}/10")
print("=" * 50)

Education Score : 10/10


In [63]:
experience_keywords = [
    "intern",
    "internship",
    "experience",
    "worked",
    "employee",
    "project"
]

experience_score = 0

for keyword in experience_keywords:
    if keyword in resume_lower:
        experience_score = 10
        break

print("=" * 50)
print(f"Experience Score : {experience_score}/10")
print("=" * 50)

Experience Score : 10/10


In [64]:
# Checking whether the important resume sections are present

section_keywords = {
    "About": ["about", "aboutme"],
    "Technical Skills": ["technical skills", "technicalskills"],
    "Education": ["education"],
    "Projects": ["projects"],
    "Experience": ["experience", "workexperience", "intern"],
    "Certifications": ["certifications", "certification", "training"]
}

text = resume_text.lower()
count = 0

for keywords in section_keywords.values():
    for word in keywords:
        if word in text:
            count += 1
            break      # Count each section only once

score = (count / len(section_keywords)) * 10

print("-" * 50)
print("Sections Found:", count, "/", len(section_keywords))
print("Resume Completeness Score:", round(score, 2), "/10")
print("-" * 50)

--------------------------------------------------
Sections Found: 6 / 6
Resume Completeness Score: 10.0 /10
--------------------------------------------------


In [65]:
total_score = (
    semantic_score +
    skills_score +
    education_score +
    experience_score +
    completeness_score
)

print("-" * 60)
print("Final ATS Score :", round(total_score, 2), "/100")
print("-" * 60)

------------------------------------------------------------
Final ATS Score : 74.2 /100
------------------------------------------------------------


# Phase 11 – Build the Streamlit Web Application


In [66]:
import os

os.chdir(r"C:\Users\user\Desktop\AI_Resume_Analyzer")

print(os.getcwd())

C:\Users\user\Desktop\AI_Resume_Analyzer


In [67]:
from utils import *

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

In [68]:
# Testing the complete pipeline

print(extract_text_from_pdf)
print(preprocess_text)
print(calculate_skill_match_score)
print(calculate_completeness_score)
print(calculate_ats_score)
print(generate_ai_suggestions)

<function extract_text_from_pdf at 0x000001DF61882F20>
<function preprocess_text at 0x000001DF619CDD00>
<function calculate_skill_match_score at 0x000001DF6CD8AE80>
<function calculate_completeness_score at 0x000001DF6CD8AFC0>
<function calculate_ats_score at 0x000001DF6CD8AF20>
<function generate_ai_suggestions at 0x000001DF6CD8ADE0>


In [69]:
# test the actual resume 

resume_path = "sample_resume.pdf"

resume_text = extract_text_from_pdf(resume_path)

print("Resume text extracted successfully.")
print("Characters extracted:", len(resume_text))
print()
print(resume_text[:1000])

Resume text extracted successfully.
Characters extracted: 4315

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI)

In [70]:
import os

print("Current folder:")
print(os.getcwd())

print("\nFiles in this folder:")
print(os.listdir())

Current folder:
C:\Users\user\Desktop\AI_Resume_Analyzer

Files in this folder:
['.ipynb_checkpoints', '.vscode', 'app.py', 'cd', 'New Text Document.txt', 'README.md', 'requirements.txt', 'Resume_Analyzer.ipynb', 'sample_jd.txt', 'sample_resume.pdf', 'utils.py', '__pycache__']


In [71]:
import os

pdf_path = os.path.abspath("sample_resume.pdf")

print(pdf_path)
print("Exists:", os.path.exists(pdf_path))
print("Is file:", os.path.isfile(pdf_path))
print("Size:", os.path.getsize(pdf_path), "bytes")

C:\Users\user\Desktop\AI_Resume_Analyzer\sample_resume.pdf
Exists: True
Is file: True
Size: 40980 bytes


In [72]:
resume_text = extract_text_from_pdf("sample_resume.pdf")

print("Resume text extracted successfully.")
print("Characters extracted:", len(resume_text))
print()
print(resume_text[:1500])

Resume text extracted successfully.
Characters extracted: 4315

GOPAL YADAV
Mumbai,Maharashtra,India | +91-7400124757 | gopalyadav18832@gmail.com
linkedin.com/in/gopal-yadav-b637b73a0 | github.com/gopalyadav-ai
ABOUTME
Final-yearB.Sc. ComputerSciencestudentspecializinginMachineLearning,DeepLearning,NLP,andGenerativeAI,withthree
end-to-endAIprojectsindependentlybuilt,evaluated,anddeployed. Deliveredmeasurablemodelperformance,includinga98.08%
accuracyhealthcareriskmodelandan80%accuracysentimentanalysismodel. ExperiencedacrossthefullMLlifecycle–data
preprocessing,exploratorydataanalysis,featureengineering,modelbuilding,hyperparametertuning,evaluation,anddeployment–
usingPython,Scikit-learn,TensorFlow/Keras,LangChain,andIBMWatsonx. Completed8industrycertificationsfromIBM,AWS,and
Microsoft/LinkedIncoveringmachinelearning,dataanalysis,SQL,andGenerativeAI,reflectingconsistentself-directedupskilling.
TechnicalSkills
Programming&QueryLanguages: Python,SQL
CoreDomains: ArtificialIntelligence(AI)

In [73]:
clean_resume_text = preprocess_text(resume_text)

print("Preprocessing completed.")
print("Characters after preprocessing:", len(clean_resume_text))
print()
print(clean_resume_text[:1500])

Preprocessing completed.
Characters after preprocessing: 4233

gopal yadav mumbai maharashtra india 91 7400124757 gopalyadav18832 gmail com linkedin com in gopal yadav b637b73a0 github com gopalyadav ai aboutme final yearb sc computersciencestudentspecializinginmachinelearning deeplearning nlp andgenerativeai withthree end to endaiprojectsindependentlybuilt evaluated anddeployed deliveredmeasurablemodelperformance includinga98 08 accuracyhealthcareriskmodelandan80 accuracysentimentanalysismodel experiencedacrossthefullmllifecycle data preprocessing exploratorydataanalysis featureengineering modelbuilding hyperparametertuning evaluation anddeployment usingpython scikit learn tensorflow keras langchain andibmwatsonx completed8industrycertificationsfromibm aws and microsoft linkedincoveringmachinelearning dataanalysis sql andgenerativeai reflectingconsistentself directedupskilling technicalskills programming querylanguages python sql coredomains artificialintelligence ai machinelearning m

In [74]:
resume_skills = extract_skills(clean_resume_text)

print("Resume skills:")
print(resume_skills)

print("\nNumber of skills found:", len(resume_skills))

Resume skills:
['aws', 'computer vision', 'git', 'github', 'keras', 'langchain', 'llm', 'llms', 'nlp', 'numpy', 'pandas', 'python', 'rag', 'sql', 'streamlit', 'tensorflow']

Number of skills found: 16


In [75]:
with open("sample_jd.txt", "r", encoding="utf-8") as file:
    jd_text = file.read()

print("Job Description loaded successfully.")
print("Characters:", len(jd_text))
print()
print(jd_text[:1500])

Job Description loaded successfully.
Characters: 1693

AI/ML Engineer Intern

Job Description:

We are looking for an enthusiastic AI/ML Engineer Intern to join our team and work on real-world Artificial Intelligence and Machine Learning applications.

Responsibilities:

- Develop and implement machine learning models using Python and Scikit-learn.
- Build and train deep learning models using TensorFlow and Keras.
- Perform data preprocessing, exploratory data analysis, feature engineering, and model evaluation.
- Work with Natural Language Processing (NLP) and Large Language Models (LLMs).
- Develop Generative AI applications using prompt engineering and Retrieval-Augmented Generation (RAG).
- Work with LangChain to build AI-powered applications.
- Analyze datasets using Pandas and NumPy.
- Use SQL for data querying and data manipulation.
- Build and deploy machine learning applications using Streamlit.
- Collaborate with the team to improve model performance and experiment with diffe

In [76]:
import os

jd_path = os.path.abspath("sample_jd.txt")

print(jd_path)
print("Exists:", os.path.exists(jd_path))
print("Is file:", os.path.isfile(jd_path))
print("Is directory:", os.path.isdir(jd_path))

if os.path.exists(jd_path) and os.path.isfile(jd_path):
    print("Size:", os.path.getsize(jd_path), "bytes")

C:\Users\user\Desktop\AI_Resume_Analyzer\sample_jd.txt
Exists: True
Is file: True
Is directory: False
Size: 1750 bytes


In [77]:
with open("sample_jd.txt", "r", encoding="utf-8") as file:
    jd_text = file.read()

print("Job Description loaded successfully.")
print("Characters:", len(jd_text))
print()
print(jd_text[:1500])

Job Description loaded successfully.
Characters: 1693

AI/ML Engineer Intern

Job Description:

We are looking for an enthusiastic AI/ML Engineer Intern to join our team and work on real-world Artificial Intelligence and Machine Learning applications.

Responsibilities:

- Develop and implement machine learning models using Python and Scikit-learn.
- Build and train deep learning models using TensorFlow and Keras.
- Perform data preprocessing, exploratory data analysis, feature engineering, and model evaluation.
- Work with Natural Language Processing (NLP) and Large Language Models (LLMs).
- Develop Generative AI applications using prompt engineering and Retrieval-Augmented Generation (RAG).
- Work with LangChain to build AI-powered applications.
- Analyze datasets using Pandas and NumPy.
- Use SQL for data querying and data manipulation.
- Build and deploy machine learning applications using Streamlit.
- Collaborate with the team to improve model performance and experiment with diffe

In [78]:
clean_jd_text = preprocess_text(jd_text)

print("JD preprocessing completed.")
print("Characters after preprocessing:", len(clean_jd_text))
print()
print(clean_jd_text[:1500])

JD preprocessing completed.
Characters after preprocessing: 1570

ai ml engineer intern job description we are looking for an enthusiastic ai ml engineer intern to join our team and work on real world artificial intelligence and machine learning applications responsibilities develop and implement machine learning models using python and scikit learn build and train deep learning models using tensorflow and keras perform data preprocessing exploratory data analysis feature engineering and model evaluation work with natural language processing nlp and large language models llms develop generative ai applications using prompt engineering and retrieval augmented generation rag work with langchain to build ai powered applications analyze datasets using pandas and numpy use sql for data querying and data manipulation build and deploy machine learning applications using streamlit collaborate with the team to improve model performance and experiment with different machine learning algorithms u

In [79]:
jd_skills = extract_skills(clean_jd_text)

print("JD skills:")
print(jd_skills)

print("\nNumber of JD skills found:", len(jd_skills))

JD skills:
['aws', 'computer vision', 'deep learning', 'git', 'github', 'keras', 'langchain', 'llm', 'llms', 'machine learning', 'natural language processing', 'nlp', 'numpy', 'pandas', 'prompt engineering', 'python', 'rag', 'sql', 'streamlit', 'tensorflow']

Number of JD skills found: 20


In [80]:
matched_skills = semantic_skill_matching(
    clean_resume_text,
    jd_skills
)

print("Matched skills:")
print(matched_skills)

print("\nNumber of matched skills:", len(matched_skills))

Matched skills:
{}

Number of matched skills: 0


In [81]:
skill_match_score = calculate_skill_match_score(
    matched_skills,
    jd_skills
)

print("Skill Match Score:", skill_match_score)

Skill Match Score: 0.0


In [82]:
matched_skills = sorted(set(resume_skills) & set(jd_skills))
missing_skills = sorted(set(jd_skills) - set(resume_skills))

print("Matched skills:")
print(matched_skills)

print("\nMissing skills:")
print(missing_skills)

print("\nMatched:", len(matched_skills))
print("Missing:", len(missing_skills))

Matched skills:
['aws', 'computer vision', 'git', 'github', 'keras', 'langchain', 'llm', 'llms', 'nlp', 'numpy', 'pandas', 'python', 'rag', 'sql', 'streamlit', 'tensorflow']

Missing skills:
['deep learning', 'machine learning', 'natural language processing', 'prompt engineering']

Matched: 16
Missing: 4


In [83]:
for skill in ["machine learning", "deep learning",
              "natural language processing", "prompt engineering"]:
    print(skill, "->", skill in resume_text.lower())

machine learning -> False
deep learning -> False
natural language processing -> False
prompt engineering -> False


In [84]:
similarity_score = calculate_similarity(
    clean_resume_text,
    clean_jd_text
)

print("Semantic Similarity Score:", similarity_score)
print("Semantic Similarity Percentage:", similarity_score * 100)

Semantic Similarity Score: 0.6068449020385742
Semantic Similarity Percentage: 60.68449020385742


In [85]:
completeness_score, sections_found = calculate_completeness_score(
    clean_resume_text
)

print("Completeness Score:", completeness_score)
print("Sections Found:", sections_found)

Completeness Score: 10.0
Sections Found: 6


In [86]:
import py_compile

py_compile.compile("utils.py", doraise=True)

print("utils.py syntax is correct.")

utils.py syntax is correct.


In [87]:
import importlib
import utils

importlib.reload(utils)

print("utils.py reloaded successfully.")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

utils.py reloaded successfully.


In [88]:
ats_score = calculate_ats_score(
    similarity_score,
    skill_match_score,
    completeness_score
)

print("Final ATS Score:", round(ats_score, 2))

Final ATS Score: 34.27


In [89]:
print("similarity_score =", similarity_score)
print("skill_match_score =", skill_match_score)
print("completeness_score =", completeness_score)

similarity_score = 0.6068449020385742
skill_match_score = 0.0
completeness_score = 10.0


In [90]:
matched_skills = sorted(set(resume_skills) & set(jd_skills))

missing_skills = sorted(set(jd_skills) - set(resume_skills))

skill_match_score = calculate_skill_match_score(
    matched_skills,
    jd_skills
)

print("Matched skills:", matched_skills)
print("Missing skills:", missing_skills)
print("Skill Match Score:", skill_match_score)

Matched skills: ['aws', 'computer vision', 'git', 'github', 'keras', 'langchain', 'llm', 'llms', 'nlp', 'numpy', 'pandas', 'python', 'rag', 'sql', 'streamlit', 'tensorflow']
Missing skills: ['deep learning', 'machine learning', 'natural language processing', 'prompt engineering']
Skill Match Score: 80.0


In [91]:
from utils import calculate_ats_score

In [92]:
ats_score = calculate_ats_score(
    similarity_score,
    skill_match_score,
    completeness_score
)

print("Final ATS Score:", round(ats_score, 2))

Final ATS Score: 74.27


In [93]:
suggestions = generate_ai_suggestions(
    matched_skills,
    missing_skills,
    ats_score,
    similarity_score,
    completeness_score
)

print("AI Resume Suggestions:\n")

for i, suggestion in enumerate(suggestions, 1):
    print(f"{i}. {suggestion}")

AI Resume Suggestions:

1. Your ATS score is moderate. Adding more relevant keywords and improving resume structure can increase your score.
2. Consider adding these missing skills if you genuinely have experience with them: deep learning, machine learning, natural language processing, prompt engineering
3. Your resume has good semantic alignment with the job description.
4. Your resume contains most of the important standard sections.


In [94]:
import os

print(os.getcwd())
print(os.listdir())

C:\Users\user\Desktop\AI_Resume_Analyzer
[]


In [95]:
import os

print(os.path.abspath("AI_Resume_Analyzer_.ipynb"))

C:\Users\user\Desktop\AI_Resume_Analyzer\AI_Resume_Analyzer_.ipynb
